File: src/main/java/com/bajdcc/LALR1/grammar/error/LostHandler.java
package com.bajdcc.LALR1.grammar.error;

import com.bajdcc.LALR1.syntax.handler.IErrorHandler;
import com.bajdcc.util.TrackerErrorBag;
import com.bajdcc.util.lexer.regex.IRegexStringIterator;

/**
 * 【语法分析】应对丢失符号的错误处理器
 *
 * @author bajdcc
 */
public class LostHandler implements IErrorHandler {

	/**
	 * 错误信息
	 */
	private String message;

	public LostHandler(String message) {
		this.message = message;
	}

	@Override
	public String handle(IRegexStringIterator iterator, TrackerErrorBag bag) {
		bag.setRead(false);
		bag.setPass(true);
		String def = "缺少：" + message + System.lineSeparator();
		String snapshot = iterator.ex().getErrorSnapshot(bag.getPosition());
		if (snapshot != null) {
			return def + snapshot;
		}
		return def;
	}
}

File: src/main/java/com/bajdcc/LALR1/grammar/error/LostHandler.kt
package com.bajdcc.LALR1.grammar.error

import com.bajdcc.LALR1.syntax.handler.IErrorHandler
import com.bajdcc.util.TrackerErrorBag
import com.bajdcc.util.lexer.regex.IRegexStringIterator

/**
 * 【语法分析】应对丢失符号的错误处理器
 *
 * @author bajdcc
 */
class LostHandler(private val message: String) : IErrorHandler {

    override fun handle(iterator: IRegexStringIterator, bag: TrackerErrorBag): String {
        bag.read = false
        bag.pass = true
        val def = "缺少：" + message + System.lineSeparator()
        val snapshot = iterator.ex().getErrorSnapshot(bag.position)
        return if (!snapshot.isEmpty()) {
            def + snapshot
        } else def
    }
}
